src/main/java/com/demo/folder/service/TrainerService.java

```diff
@@ -1,108 +1,347 @@
+    }
+    LOGGER.info("Creating new Trainer with first name: {}", trainerRequestDTO.getFirstName());
+    Optional<TrainingType> existingTrainingType = trainingTypeRepository.findById(
+        trainerRequestDTO.getTrainingTypeId());
+    String[] arr = new String[2];
+    String plainTextPassword = Generator.generatePassword();
+    arr[0] = plainTextPassword;
+
+    TrainingType trainingType;
+    if (existingTrainingType.isPresent()) {
+      // If the TrainingType exists, use the existing one
+      trainingType = existingTrainingType.get();
+    } else {
+      return null;
+    }
+
+    // Create User entity for the Trainer
+    User user = new User();
+    user.setFirstName(trainerRequestDTO.getFirstName());
+    user.setLastName(trainerRequestDTO.getLastName());
+    user.setUsername(Generator.generateUserName(trainerRequestDTO.getFirstName(),
+        trainerRequestDTO.getLastName()));
+    user.setPassword(passwordEncoder.encode(plainTextPassword));
+    user.setActive(true);
+
+    // Create Trainer entity
+    Trainer trainer = new Trainer();
+    trainer.setSpecialization(trainingType);
+    trainer.setUser(user);
+
+    // Save the Trainer
+    trainerRepository.save(trainer);
+
+    arr[1] = user.getUsername();
+
+    return arr;
+  }
+
+
+  @Transactional(readOnly = true)
+  public Trainer findTrainerByUsername(String username) {
+    LOGGER.info("Finding Trainer by username: {}", username);
+    if (username == null || username.isEmpty()) {
+      LOGGER.info("Username cannot be null or empty.");
+    }
+
+    Trainer trainer = trainerRepository.findByUsername(username);
+    if (trainer == null) {
+      LOGGER.info("Trainer with username {} not found.", username);
+    }
+    return trainer;
   }
 
+  @Transactional
+  public void activateTrainer(Long trainerId) {
+    LOGGER.info("Activating Trainer with ID: {}", trainerId);
+    trainerRepository.updateTrainerStatus(trainerId, true);
+  }
+
+  @Transactional
+  public void deactivateTrainer(Long trainerId) {
+    LOGGER.info("Deactivating Trainer with ID: {}", trainerId);
+    trainerRepository.updateTrainerStatus(trainerId, false);
+  }
+
+  @Transactional(readOnly = true)
   public List<Trainer> getAllTrainers() {
-    return trainerDAO.getAll();
+    LOGGER.info("Fetching all Trainers.");
+    List<Trainer> trainers = trainerRepository.findAll();
+
+    if (trainers.isEmpty()) {
+      throw new EntityNotFoundException("No trainers found.");
+    }
+    return trainers;
+  }
+
+  @Transactional
+  public void updateTrainer(Trainer trainer) {
+    String username = trainer.getUser().getUsername();
+    LOGGER.info("Updating Trainer with username: {}", trainer.getUser().getUsername());
+    Trainer existingTrainer = trainerRepository.findByUsername(username);
+    if (existingTrainer == null) {
+      throw new EntityNotFoundException("Trainer with username " + username + " not found.");
+    }
```

Comment:
I believe this check is unnecessary here, as in the places where this method is called a similar check is already made. 
Also, consider using the Transactional annotation, this method is marked with it, and the method in which it is called is also marked with this annotation, what will happen in this case?